File: A2/A2-020/A2-020.cpp
#include<iostream>
#include<vector>
using namespace std;

int main()
{
    // This makes std::cin and std::cout faster
    ios_base::sync_with_stdio(false); cin.tie(0);

    // Input amount of people
    int N;
    cin >> N;
    // Input graph of items flows
    vector<int> graph(N, -1);
    for(int from = 0; from < N; from++)
    {
        int to;
        cin >> to;
        graph[from] = to - 1;
    }

    // Find maximum node amount on cycle on a graph
    // Every node has 1 in-degree and 1 out-degree
    int maxCycleCount = 0;
    vector<bool> isVisited(N, false);
    for(int start = 0; start < N; start++)
    {
        // Find start node of a cycle
        if(isVisited[start]) { continue; }
        
        // Count node on a cycle
        int curr = start;
        int cycleCount = 0;
        while(!isVisited[curr])
        {
            isVisited[curr] = true;
            curr = graph[curr];
            cycleCount++;
        }
        // Find maximum node amount on a cycle
        maxCycleCount = max(maxCycleCount, cycleCount);
    }

    // Output maximum node amount on cycle on a graph
    cout << maxCycleCount;
    return 0;
}

File: A2/A2-020/A2-020.py
# Input amount of people
N = int(input())

# Input graph of items flows
graph = [-1] * N
for node_from in range(N):
    node_to = int(input())
    graph[node_from] = node_to - 1

# Find maximum node amount on cycle on a graph
# Every node has 1 in-degree and 1 out-degree
max_cycle_count = 0
is_visited = [False] * N
for start in range(N):
    # Find start node of a cycle
    if(is_visited[start]):
        continue
    
    # Count node on a cycle
    curr = start
    cycle_count = 0
    while(not is_visited[curr]):
        is_visited[curr] = True
        curr = graph[curr]
        cycle_count += 1
    # Find maximum node amount on a cycle
    max_cycle_count = max(max_cycle_count, cycle_count)

# Output maximum node amount on cycle on a graph
print(max_cycle_count)